### ISSUE TITLE
HoverProvider appears in the wrong position

### ISSUE THREAD

**Issue by liby** [rust-analyzer/rust-analyzer#9974]
https://user-images.githubusercontent.com/38807139/130327270-119ca235-5650-4aad-a046-b47827d4703b.mp4

[Example code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8028f03b6efbc72d77e16e330fd7759e)

**Comment by Veykril** [comment]
That is due to the diagnostic being split apart so what you are actually hovering is the diagnostic which vscode shows at the appropriate(primary) location.

**Comment by liby** [comment]
@Veykril 
> That is due to the diagnostic being split apart
I don't quite understand the meaning of this sentence. Can you explain it in detail?


https://user-images.githubusercontent.com/38807139/130328249-6dbfab5a-abb6-4385-a304-f2da178f8df6.mp4

**Comment by Veykril** [comment]
Basically the `Screen` identifier you are hovering has an error span below it(since its unresolved in this case). Rust also emits another error span in lines 15-19, this error span is related to the first one and as such vscode moves the pop up down to the lines below because they also belong to the error.

And due to that vscode moves the entire hover box down because it merges the normal hover message with the error message.

In short this is expected behaviour at least for vscode.

**Comment by liby** [comment]
@Veykril 
> Basically the Screen identifier you are hovering has an error span below it(since its unresolved in this case). Rust also emits another error span in lines 15-19, this error span is related to the first one and as such vscode moves the pop up down to the lines below because they also belong to the error.

After I removed the code from lines 15-19, this issue still exists. 

> And due to that vscode moves the entire hover box down because it merges the normal hover message with the error message.

It looks like it flashes to the top once in a while.

https://user-images.githubusercontent.com/38807139/130328851-7721ad04-d68b-44ee-afe3-718eb0d5794a.mp4

**Comment by Veykril** [comment]
That flashing happens because the error also points to the top of the file, note the `main.rs(1, 1)` in the message, so vscode tries to move the diagnostic hover to the top first and then puts it back down for some reason.

The main problem here is VScode not handling multi span diagnostics too well which we create from rustc's output as it likes to point over the entire file with its diagnostics.

**Comment by liby** [comment]
> The main problem here is VScode not handling multi span diagnostics too well which we create from rustc's output as it likes to point over the entire file with its diagnostics.

I am not very accepting of this explanation. For example, the following screenshot. The range of errors reported is from 8 to 15 lines, but the diagnosis always appears near the pointer hover rather than at the bottom of the whole block.

https://user-images.githubusercontent.com/38807139/130329155-f5ed2dde-df91-4be8-b12b-bb5d40f3d386.mp4

**Comment by Veykril** [comment]
Rust Analyzer is not responsible for the position of the hover window, if this positioning seems odd to you you should open an issue in the vscode repo instead.

**Comment by liby** [comment]
@Veykril 
Sorry to interrupt. Following your suggestion, I have created a [issue](https://github.com/microsoft/vscode/issues/131326) in the VSCode repo. In that issue, @gjsjohnmurray mentioned that:
> Originally posted by @Veykril in rust-analyzer/rust-analyzer#9974 (comment)

> @Veykril it may help the VS Code devs understand this issue if you are able to post the JSON of the Diagnostic object(s) that Rust Analyzer will be returning in the example given by @liby

I don't know if you have seen it, but would it be convenient for you to provide it to advance the progress of the issue? Or is there anything I can do to help?

**Comment by Veykril** [comment]
Since you have a small reproduction snippet you can collect the diagnostic objects yourself I believe. For that you only need to set the following config `"rust-analyzer.trace.server": "verbose"`, then save the file(assuming you have check on save enabled) so that RA emits diagnostics. Then you can check the `Rust Analyzer Language Server Trace` logs and take the json object received in the `Received notification 'textDocument/publishDiagnostics'` part.
That is this notification here is what the VSCode issue wants 
![image](https://user-images.githubusercontent.com/3757771/130364170-369f7815-4d59-4704-bd57-3e9876961cd7.png)
